File: frontend/src/features/subject/types/subject.ts
export default interface Subject {
  id: string;
  name: string;
  code: string;
  credits: number;
  description?: string;
  faculty?: string;
  prerequisites?: Partial<Subject>[];
}

export interface CreateSubjectDTO {
  name: string;
  code: string;
  credits: number;
  description?: string;
  faculty?: string;
  prerequisites?: string[];
}

export interface UpdateSubjectDTO {
  name?: string;
  code?: string;
  credits?: number;
  description?: string;
  faculty?: string;
  prerequisites?: string[];
}

export const mapToSubject = (data: any): Subject => ({
  id: data.id as string,
  name: data.name,
  code: data.code,
  credits: data.credits,
  description: data.description,
  faculty: data.faculty,
  prerequisites: data.prerequisitesSubjects,
});

File: backend/src/main/java/com/tkpm/sms/domain/model/Subject.java
package com.tkpm.sms.domain.model;

import com.tkpm.sms.domain.utils.TranslationUtils;
import com.tkpm.sms.domain.valueobject.TextContent;
import lombok.*;
import lombok.experimental.FieldDefaults;

import java.time.LocalDateTime;
import java.util.List;

@Getter
@Setter
@AllArgsConstructor
@NoArgsConstructor
@Builder
@FieldDefaults(level = AccessLevel.PRIVATE)
public class Subject {
    Integer id;
    TextContent name;
    String code;
    boolean isActive;
    TextContent description;
    LocalDateTime createdAt;
    Integer credits;
    Faculty faculty;
    List<Subject> prerequisites;

    public String getNameByLanguage(String language) {
        return TranslationUtils.getTextByLanguage(name, language);
    }

    public String getDescriptionByLanguage(String language) {
        return TranslationUtils.getTextByLanguage(description, language);
    }
}